device/receive.go

```diff
@@ -103,100 +117,117 @@ func (device *Device) RoutineReceiveIncoming(recv conn.ReceiveFunc) {
 				deathSpiral++
 				time.Sleep(time.Second / 3)
-				buffer = device.GetMessageBuffer()
 				continue
 			}
 			return
 		}
 		deathSpiral = 0
 
-		if size < MinMessageSize {
-			continue
-		}
+		// handle each packet in the batch
+		for i, size := range sizes[:count] {
+			if size < MinMessageSize {
+				// this is also deliberately skipping len 0
+				continue
+			}
 
-		// check size of packet
+			// check size of packet
 
-		packet := buffer[:size]
-		msgType := binary.LittleEndian.Uint32(packet[:4])
+			packet := buffsArrs[i][:size]
+			msgType := binary.LittleEndian.Uint32(packet[:4])
 
-		var okay bool
+			var okay bool
 
-		switch msgType {
+			switch msgType {
 
-		// check if transport
+			// check if transport
 
-		case MessageTransportType:
+			case MessageTransportType:
 
-			// check size
+				// check size
 
-			if len(packet) < MessageTransportSize {
-				continue
-			}
+				if len(packet) < MessageTransportSize {
+					continue
+				}
 
-			// lookup key pair
+				// lookup key pair
 
-			receiver := binary.LittleEndian.Uint32(
-				packet[MessageTransportOffsetReceiver:MessageTransportOffsetCounter],
-			)
-			value := device.indexTable.Lookup(receiver)
-			keypair := value.keypair
-			if keypair == nil {
-				continue
-			}
+				receiver := binary.LittleEndian.Uint32(
+					packet[MessageTransportOffsetReceiver:MessageTransportOffsetCounter],
+				)
+				value := device.indexTable.Lookup(receiver)
+				keypair := value.keypair
+				if keypair == nil {
+					continue
+				}
 
-			// check keypair expiry
+				// check keypair expiry
 
-			if keypair.created.Add(RejectAfterTime).Before(time.Now()) {
-				continue
-			}
+				if keypair.created.Add(RejectAfterTime).Before(time.Now()) {
+					continue
+				}
 
-			// create work element
-			peer := value.peer
-			elem := device.GetInboundElement()
-			elem.packet = packet
-			elem.buffer = buffer
-			elem.keypair = keypair
-			elem.endpoint = endpoint
-			elem.counter = 0
-			elem.Mutex = sync.Mutex{}
-			elem.Lock()
+				// create work element
+				peer := value.peer
+				elem := device.GetInboundElement()
+				elem.packet = packet
+				elem.buffer = buffsArrs[i]
+				elem.keypair = keypair
+				elem.endpoint = endpoints[i]
+				elem.counter = 0
+				elem.Mutex = sync.Mutex{}
+				elem.Lock()
+
+				elemsForPeer, ok := elemsByPeer[peer]
+				if !ok {
+					elemsForPeer = device.GetInboundElementsSlice()
+					elemsByPeer[peer] = elemsForPeer
+				}
+				*elemsForPeer = append(*elemsForPeer, elem)
+				buffsArrs[i] = nil
+				continue
 
-			// add to decryption queues
-			if peer.isRunning.Load() {
-				peer.queue.inbound.c <- elem
-				device.queue.decryption.c <- elem
-				buffer = device.GetMessageBuffer()
-			} else {
-				device.PutInboundElement(elem)
-			}
-			continue
+			// otherwise it is a fixed size & handshake related packet
 
-		// otherwise it is a fixed size & handshake related packet
+			case MessageInitiationType:
+				okay = len(packet) == MessageInitiationSize
```

Comment:
Wouldn't it make more sense to just `continue` if len!=expected and inside of default, rather than the `okay` boolean? Seems like the restructuring of the function lets that be simplified as such now.

Reply:
Updated in https://github.com/WireGuard/wireguard-go/pull/64/commits/819739aab87a51b24ba7faf58c511e57ef9c2525.